js/modules/loki_snode_api.js

```diff
@@ -35,77 +36,122 @@ class LokiSnodeAPI {
-            consecutiveErrors
+            'loki_snodes: Refreshed random snode pool with',
+            this.randomSnodePool.length,
+            'snodes'
           );
-          this.initialiseRandomPool(seedNodes, consecutiveErrors + 1);
-        }, consecutiveErrors * consecutiveErrors * 5000);
-      } else {
-        log.error('Giving up trying to contact seed node');
-        if (snodes.length === 0) {
-          throw new window.textsecure.SeedNodeError(
-            'Failed to contact seed node'
+        } catch (e) {
+          log.warn(
+            'loki_snodes: initialiseRandomPool error',
+            e.code,
+            e.message
           );
+          if (consecutiveErrors < 3) {
+            // retry after a possible delay
+            setTimeout(() => {
+              log.info(
+                'loki_snodes: Retrying initialising random snode pool, try #',
+                consecutiveErrors
+              );
+              this.initialiseRandomPool(seedNodes, consecutiveErrors + 1);
+            }, consecutiveErrors * consecutiveErrors * 5000);
+          } else {
+            log.error('loki_snodes: Giving up trying to contact seed node');
+            if (snodes.length === 0) {
+              throw new window.textsecure.SeedNodeError(
+                'Failed to contact seed node'
+              );
+            }
+          }
         }
-      }
+        // clear lock
+        this.initialiseRandomPoolPromise = null;
+        resolve();
+      });
     }
+    await this.initialiseRandomPoolPromise;
   }
 
-  // nodeUrl is like 9hrje1bymy7hu6nmtjme9idyu3rm8gr3mkstakjyuw1997t7w4ny.snode
-  async unreachableNode(pubKey, nodeUrl) {
+  // unreachableNode.url is like 9hrje1bymy7hu6nmtjme9idyu3rm8gr3mkstakjyuw1997t7w4ny.snode
+  async unreachableNode(pubKey, unreachableNode) {
     const conversation = ConversationController.get(pubKey);
     const swarmNodes = [...conversation.get('swarmNodes')];
-    const filteredNodes = swarmNodes.filter(
-      node => node.address !== nodeUrl && node.ip !== nodeUrl
-    );
+    if (typeof unreachableNode === 'string') {
+      log.warn(
+        'loki_snodes::unreachableNode: String passed as unreachableNode to unreachableNode'
+      );
+      return swarmNodes;
+    }
+    let found = false;
+    const filteredNodes = swarmNodes.filter(node => {
+      // keep all but thisNode
+      const thisNode =
+        node.address === unreachableNode.address &&
+        node.ip === unreachableNode.ip &&
+        node.port === unreachableNode.port;
+      if (thisNode) {
+        found = true;
+      }
+      return !thisNode;
+    });
+    if (!found) {
+      log.warn(
+        `loki_snodes::unreachableNode snode ${unreachableNode.ip}:${
+          unreachableNode.port
+        } has already been marked as bad`
+      );
+    }
     await conversation.updateSwarmNodes(filteredNodes);
+    return filteredNodes;
   }
 
   markRandomNodeUnreachable(snode) {
```

Comment:
The name doesn't seem to suggest that it would return something. I see that `randomSnodePool.lenght` is only used for logging, why not log inside `markRandomNodeUnreachable`? Like "Marked {} as unreachable, {} left in the pool". Then you can have another log message one level above telling the reason (which can be different from what I can see).

Reply:
it mainly to reduce the number of logging messages. I had that way initially but we already log everytime this is called as it's a p. big deal event.

Reply:
If you don't want two separate log messages, how about we make another call to `getRandomSnodePool()` to get the size?

Reply:
well both actions need to happening the banning of the node and getting the remaining count. Is it the name of the function that is the issue?

A compromise maybe the two function solution but seems a bit inefficient, since banning the node needs to happen and knows the count. But I guess it is just a member variable it needs to count, so not as bad as the `unreachableNode` node situation.

I'll just split for now. Due to the time crunch of trying to get this in for 1.0.5, I'll leave a temp variable in for now and clean it up later.